File: server/datasets/python_source-code/Multiple-C2/unique_1.py
"""
0/1 Knapsack (DP) - Unique Submission
A genuinely distinct approach: branch-and-bound backtracking with pruning
via a fractional-knapsack upper bound, instead of a bottom-up/top-down DP
table. Still solves the same 0/1 knapsack problem exactly.
"""


def bound(index, current_weight, current_value, weights, values, capacity):
    """Upper bound on achievable value from this state, using the
    fractional relaxation of the remaining items (sorted by value/weight)."""
    if current_weight >= capacity:
        return 0

    total_value = current_value
    remaining_capacity = capacity - current_weight

    for i in range(index, len(weights)):
        if weights[i] <= remaining_capacity:
            remaining_capacity -= weights[i]
            total_value += values[i]
        else:
            total_value += values[i] * (remaining_capacity / weights[i])
            break

    return total_value


def knapsack_branch_and_bound(weights, values, capacity):
    # sort items by value-density descending; keeps bounding effective
    order = sorted(range(len(weights)), key=lambda i: values[i] / weights[i], reverse=True)
    sorted_weights = [weights[i] for i in order]
    sorted_values = [values[i] for i in order]

    best_value = 0
    n = len(weights)

    def explore(index, current_weight, current_value):
        nonlocal best_value

        if current_weight <= capacity and current_value > best_value:
            best_value = current_value

        if index == n:
            return

        # prune this branch if even the optimistic bound can't beat the best found
        if bound(index, current_weight, current_value, sorted_weights, sorted_values, capacity) <= best_value:
            return

        # branch: include current item
        if current_weight + sorted_weights[index] <= capacity:
            explore(index + 1, current_weight + sorted_weights[index], current_value + sorted_values[index])

        # branch: exclude current item
        explore(index + 1, current_weight, current_value)

    explore(0, 0, 0)
    return best_value


if __name__ == "__main__":
    weights = [2, 3, 4, 5]
    values = [3, 4, 5, 6]
    capacity = 5
    print(knapsack_branch_and_bound(weights, values, capacity))

File: server/datasets/java_source-code/Multiple-C2/unique_1.java
/*
 * 0/1 Knapsack (DP) - Unique Submission
 * Distinct approach: branch-and-bound backtracking with upper bound pruning.
 */

import java.util.Arrays;
import java.util.Comparator;

class Solution {
    private static double bound(int index, int currentWeight, int currentValue,
                                Integer[] order, int[] weights, int[] values, int capacity) {
        if (currentWeight >= capacity) {
            return 0;
        }
        double totalValue = currentValue;
        int remainingCapacity = capacity - currentWeight;

        for (int i = index; i < order.length; i++) {
            int origIdx = order[i];
            if (weights[origIdx] <= remainingCapacity) {
                remainingCapacity -= weights[origIdx];
                totalValue += values[origIdx];
            } else {
                totalValue += values[origIdx] * ((double) remainingCapacity / weights[origIdx]);
                break;
            }
        }
        return totalValue;
    }

    public static int knapsackBranchAndBound(int[] weights, int[] values, int capacity) {
        int n = weights.length;
        Integer[] order = new Integer[n];
        for (int i = 0; i < n; i++) order[i] = i;

        Arrays.sort(order, Comparator.comparingDouble((Integer i) -> (double) values[i] / weights[i]).reversed());

        int[] bestValue = new int[]{0};
        explore(0, 0, 0, order, weights, values, capacity, bestValue);
        return bestValue[0];
    }

    private static void explore(int index, int currentWeight, int currentValue,
                                Integer[] order, int[] weights, int[] values, int capacity, int[] bestValue) {
        if (currentWeight <= capacity && currentValue > bestValue[0]) {
            bestValue[0] = currentValue;
        }
        if (index == order.length) {
            return;
        }

        if (bound(index, currentWeight, currentValue, order, weights, values, capacity) <= bestValue[0]) {
            return;
        }

        int origIdx = order[index];
        if (currentWeight + weights[origIdx] <= capacity) {
            explore(index + 1, currentWeight + weights[origIdx], currentValue + values[origIdx],
                    order, weights, values, capacity, bestValue);
        }
        explore(index + 1, currentWeight, currentValue, order, weights, values, capacity, bestValue);
    }

    public static void main(String[] args) {
        int[] weights = {2, 3, 4, 5};
        int[] values = {3, 4, 5, 6};
        int capacity = 5;
        System.out.println(knapsackBranchAndBound(weights, values, capacity));
    }
}
